packages/sitecore-jss/src/personalize/layout-personalizer.ts

```diff
@@ -0,0 +1,74 @@
+import { LayoutServiceData, ComponentRendering, HtmlElementRendering } from './../layout/models';
+
+// NULL means Hidden by this experience
+export type ComponentRenderingWithExperiences = ComponentRendering & {
+  experiences: { [name: string]: ComponentRenderingWithExperiences | null };
+};
+
+/**
+ * recursively go through all placeholders/components, check experiences nodes and replace default with object from specific experience
+ * @param {LayoutServiceData} layout Layout data
+ * @param {string} segment segmentId
+ */
+export function personalizeLayout(layout: LayoutServiceData, segment: string): void {
+  const placeholders = layout.sitecore.route?.placeholders;
+  if (Object.keys(placeholders ?? {}).length === 0) {
+    return;
+  }
+  if (placeholders) {
+    Object.keys(placeholders).forEach((placeholder) => {
+      placeholders[placeholder] = personalizePlaceholder(placeholders[placeholder], segment);
+    });
+  }
+}
+
+/**
+ * @param {Array} components components within placeholder
+ * @param {string} segment segmentId
+ */
+export function personalizePlaceholder(
+  components: Array<ComponentRendering | HtmlElementRendering>,
+  segment: string
+): Array<ComponentRendering | HtmlElementRendering> {
+  return components.map((component) =>
```

Comment:
I think you can avoid the casting to `unkown` being done if you specify the `component` type?
```suggestion
  return components.map((component: ComponentRenderingWithExperiences) =>
```

Reply:
So the component is of type `<ComponentRendering | HtmlElementRendering>` in order to pass it to the next function( `personalizeComponent`) we need to cast it into `ComponentRenderingWithExperiences` type because it only accepts  `ComponentRenderingWithExperiences` type.

Reply:
Ah, yep I missed the array type, that makes sense. Just noticed, isn't the last (falsy) value the incorrect return type? It should be `ComponentRendering | HtmlElementRendering` not `ComponentRenderingWithExperiences`, right?

Reply:
by last(false) you mean line 40?
Yes correct, it is of type `ComponentRendering | HtmlElementRendering`
I guess you saw the old file. Can you check the latest changes?

Reply:
👍 latest looks correct